### ISSUE TITLE
Using library in Angular 2

### ISSUE THREAD

**Issue by ghost** [basarat/typescript-collections#32]
Hi,

I've been ripping my hair out for hours. I think there is something that I don't seem to understand about how this whole packaging/modular stuff works in this Javascript world.

I've basically got an Angular 2 app which plays nicely with anything that has nice compatible modules already available. No matter what I try I can't get this to work. I'm also using Webpack.

I basically did the following:

`npm install typescript-collections --save`

I then did

`typings install file:./node_modules/typescript-collections/collections.d.ts`

This put the typings file in my typings folder in the "browser" and "main" folders etc and added this to my code:

`/// <reference path="browser\definitions\collections\collections.d.ts" />`

However, now I'm trying to do something like:

`this._apps = new collections.Dictionary<string, AppModel>();`

But I just simply get the error "ReferenceError: collections is not defined".

I tried:

```
import { collections } from 'typescript-collections/collections';

import 'collections';

import 'typescript-collections/collections';

require('typescript-collections');

require('../node_modules/typescript-collections/collections');
```

Still, I get the same error.

Interestingly in my Atom editor intellisense is working fine and it sees all these modules, however when running this in the browser I just get collections is not defined.

**Comment by houssemzaier** [comment]
Same problem here

**Comment by kalessine** [comment]
i use 
///<reference path="../collections.ts"/>
and have no problems here...

**Comment by kalessine** [comment]
(three slashes)<reference path="../collections.ts"/>

**Comment by kalessine** [comment]
i use 
///&gt;reference path="../collections.ts"/&lt;

**Comment by kalessine** [comment]
I use
///&lt;reference path="../collections.ts"/&gt;
and have no problems here =D

**Comment by episage** [comment]
@templar112 @houssemzaier  Please try the new version 1.1.0 

bash: `npm install typescript-collections --save`

tsconfig.json

```
{
    "compilerOptions": {
        "target": "es5",
        "module": "commonjs",
        "declaration": true,
        "noImplicitAny": true,
        "removeComments": false,
        "sourceMap": true,
        "outDir": "./dist",
        "noLib": false,
        "preserveConstEnums": true
    },
    "exclude": [
        "node_modules",
        "typings/browser",
        "typings/browser.d.ts",
        "examples"
    ]
}
```

You don't need typings for typescript-collections (ts.d files) because they are embedded into npm package and your IDE should autodiscover them.

**Comment by episage** [comment]
Reopen if you still have the issue.

**Comment by patrickracicot** [comment]
Hey,

I don't know if this is related but I've installed the library today in my Angular2 project and everything is working fine in my IDE.  
But when I try to open my app, I get the following error : 
`angular2-polyfills.min.js:1 Error: SyntaxError: Unexpected token`

After verifying my network tab in Chrome, I saw that the actual path used to evaluate the import in the ts file has 2 errors:
1. It doesn't point at the proper path in the solution structure (the path is basicly http://foo/typescript-collections instead of http://foo/node_modules/typescript-collections
2. The path used to evaluate is missing the file extension (.js)

Here is how I import the library in my component: `import * as Collections from 'typescript-collections';`

**Comment by episage** [comment]
Hi @patrickracicot , thanks for using `typescript-collections`. Could you please provide a minimal example for me to test?

btw. Have you tried using `var Collections = require('typescript-collections')` ?

**Comment by patrickracicot** [comment]
Hi @episage , your suggestion isn't working either.  After a bit of thinking, I came up with the idea of defining a path to the `umd.min.js` in my `System.config`.  This way when I use `import * as Collections from 'typescript-collections';`, Angular will acutally point to the mapped path which means it will load the `umd.min.js` file.

Here is a code sample of my System.config:

```
System.config({
            packages: {
                app: {
                    format: 'cjs',
                    defaultExtension: 'js'
                }
            },
            paths: {
                'typescript-collections': 'node_modules/typescript-collections/dist/lib/umd.min.js'
            }
        });
```

With this, when I use `import * as Collections from 'typescript-collections';` everything works as intended.

**Comment by fkolar** [comment]
I just wonder are there any news regardin this issue. I am getting the same things. I download latest angular CLI @webpack version, install this library and when I try to build ng build --dev I get following error:

When I run the test all works find just I see this warning during a build. How can I turn this off? 

I tried to exclude this my tsconfig.json but it seems does not work. 

```
ersion: webpack 2.1.0-beta.18
Time: 7661ms
              Asset       Size  Chunks             Chunk Names
     main.bundle.js    3.14 MB    0, 2  [emitted]  main
polyfills.bundle.js     227 kB    1, 2  [emitted]  polyfills
          inline.js    5.11 kB       2  [emitted]  inline
           main.map    3.76 MB    0, 2  [emitted]  main
      polyfills.map     293 kB    1, 2  [emitted]  polyfills
         inline.map    5.15 kB       2  [emitted]  inline
         index.html  729 bytes          [emitted]  
chunk    {0} main.bundle.js, main.map (main) 3.05 MB {1} [initial] [rendered]
chunk    {1} polyfills.bundle.js, polyfills.map (polyfills) 195 kB {2} [initial] [rendered]
chunk    {2} inline.js, inline.map (inline) 0 bytes [entry] [rendered]

WARNING in ./~/typescript-collections/dist/lib/umd.js
Cannot find SourceMap 'data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIm5vZGVfbW9kdWxlcy9icm93c2VyLXBhY2svX3ByZWx1ZGUuanMiLCJkaXN0L2xpYi9CU1RyZWUuanMiLCJkaXN0L2xpYi9CYWcuanMiLCJkaXN0L2xpYi9EaWN0aW9uYXJ5LmpzIiwiZGlzdC9saWIvSGVhcC5qcyIsImRpc3QvbGliL0xpbmtlZERpY3Rpb25hcnkuanMiLCJk
```

```
{
  "compilerOptions": {
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": ["es6", "dom"],
    "mapRoot": "./",
    "module": "es6",
    "moduleResolution": "node",
    "outDir": "../dist/out-tsc",
    "sourceMap": true,
    "target": "es5",
    "typeRoots": [
      "../node_modules/@types"
    ],
    "types": [
      "jasmine"
    ],
    "exclude": [
      "../node_modules/typescript-collections/dist/lib/*.d.ts"
    ]
  }
}


```

Thanks for any feedback. 

Frank

**Comment by fkolar** [comment]
I just temporary resolved this, by editing the umd.js. 
I took the inlined sourceMap out decoded it , and placed it into umd.js.map. and changed the sourceMap url inside the file. 

And the error is gone. Not sure what I really did, but it works!! ;-)

**Comment by patrickracicot** [comment]
@fkolar Hmm that weird, in the package pulled by npm I don't have a `umd.js.map`.  In my case it is normal since I don't have any .ts files for the umd.

**Comment by fkolar** [comment]
The thing I was trying to say, I had to create this file manually with a content that I extracted from uml.js and base64decoded.

**Comment by gforceg** [comment]
I have a similar problem. I've created a npm module using typescript that uses the Dictionary type. The module compiles without issue.

When I bring the module into my angular 2 app everything compiles without issue.

At runtime when I call a function in my node module that uses the Dictionary type, I get the error: "TypeError: Cannot read property 'getValue' of undefined."

I can copy the code into my project instead of using it as a node module and everything works fine. I don't know what I'm doing wrong.

**Comment by gforceg** [comment]
FIXED IT!
In my node module I had to import the Collections object. I think this has to do w/ the typescript-collections module exporting a umd.js. Throughout my code I was referring to 'Dictionary' instead of 'Collections.Dictionary'.

changed:
`import {Dictionary} from 'typescript-collections';`
to:
`import * as Collections from 'typescript-collections';`

and changed:
`static cache: Dictionary<String, Dictionary<String, IEvent>> = new Dictionary<String, Dictionary<String, IEvent>>();`
to:
`static cache: Collections.Dictionary<String, Collections.Dictionary<String, IEvent>> = new Collections.Dictionary<String, Collections.Dictionary<String, IEvent>>();`

**Comment by patrickracicot** [comment]
@gforceg Exactly, this modules exposes it's functionality though Collections.  You can see this in the main page of this repository in the [usage section](https://github.com/basarat/typescript-collections#usage)

**Comment by gforceg** [comment]
I know. I thought why import the whole thing when all I want is one type? So since it didn't break my code originally, I couldn't figure out why it broke my module.

Then I went back and decided to import it the prescribed way rather than the way I thought it "should be" imported.

I guess I should follow instructions next time. ;)

**Comment by episage** [comment]
Hi everyone

I found, probably, the most popular `Angular 2` starter pack on the web `https://github.com/AngularClass/angular2-webpack-starter` and tried to compile and run `typescript-collections` with it. 

As always: `npm i --save typescript-collections` then

I added 

``` javascript
import { DefaultDictionary } from 'typescript-collections';
```

and

``` javascript
    console.log('typescript-collections test start');
    let dict = new DefaultDictionary<string, string>(() => 'defaultValue');
    console.log(dict.getValue('kekemeke'));
    dict.setValue('foo', 'bar');
    console.log(dict.getValue('foo'));
```

to `home.component.ts` `ngOnInit` and it worked out of the box.
![image](https://cloud.githubusercontent.com/assets/2196095/18410347/39fcd912-7760-11e6-9041-f799b8cf33a9.png)

@gforceg I tried to replicate your issue. Selective import in the test above worked well.
@fkolar During the test my `webpack` in terminal stayed silent.
![image](https://cloud.githubusercontent.com/assets/2196095/18410357/d8c6873c-7760-11e6-88c1-6265133c406d.png)

I can't replicate the issues you mentioned. Could you please provide a minimal example or make a PR?

Best
episage

**Comment by NirLevanon-zz** [comment]
Something breaks in version 1.1.5. I can import 'typescript-collections' from earlier versions (1.1.4).

**Comment by episage** [comment]
@NirLevanon Thank you for the report. I published fixed version 1.1.8. 

Sorry for the inconvenience.

Best

**Comment by episage** [comment]
Closing this because of old age. If the issue persists, please let me know.
